### ISSUE TITLE
[FR] Integrate autofills and password management

### ISSUE THREAD

**Issue by edgar-vincent** [atlas-engineer/nyxt#1138]
It would be particularly handy if one could combine autofills with the password manager integration. I'm thinking of something similar to https://github.com/browserpass/browserpass-extension, which inserts logins and passwords in the correct fields at the press of a keychord. Entries are identified by their names, which match a given site's URL.

If I'm not mistaken, one cannot currently copy usernames in Nyxt, for instance, which is a bit of a drag.

**Comment by jmercouris** [comment]
Correct, and we have a autofill issue about that, but I think we should provide some sort of abstraction for this integration!

**Comment by fabian-thomas** [comment]
## Options

I see 4 options here:

### Using the existing autofill framework
One could add one autofill target for password and username respectively. Then the user could focus those fields manually and autofill the pw/username. This is already possible with the following really hacky snippet:
```
(make-autofill :name "username" :fill (lambda () (copy-username) (trivial-clipboard:text)))
```
The hacky part could easily be fixed by providing some get-username and get-password function in the APIs. This has the downside of having to manually select both username and password fields and triggering an autofill for both. The worst part about this triggering twice, is that one has to select the correct account twice. To fix this latter part a hacky solution until this is implemented nicely is:
```
(make-autofill :name "both" :fill (lambda () (copy-username) (let* ((username (trivial-clipboard:text))) (copy-password) username)))
```
Let the autofill framework fill the username, and copy the password to clipboard.

### Extending the autofill framework to support element matching
We could implement some way to specify multiple autofill targets in one and specify the HTML element that should be filled by XPath or Jspath. This would be the cleanest, but worst to maintain solution. This is what the browser extensions like bitwarden do.

### Misusing the existing autofill framework
As I understand it, currently one autofill target just fills/pastes the data that is specified into the currently selected field. That means we can't fill two fields nicely. As with the above case, we could extend the autofill framework to perform multiple fills with actions in between. I think about, fill username, then fake a tab key press, then fill password. I know this sounds hacky again, but that is what I used with qutebrowsers [`qute-pass` userscript](https://github.com/qutebrowser/qutebrowser/blob/65aa61f33fbada24e07a30fe78b69232573aa069/misc/userscripts/qute-pass) for over one year now, and it worked like a charm. I had 3 keybindings: one for username only and password only fill, and one for filling both. Like that I could fill any login form pretty easily.

### Above option but not misusing the autofill framework
Like I did in qutebrowser we could just add functions for `fill-username`, `fill-password`, and `fill-username-and-password`, where the last one would fake a tab key press in between calling the two other functions. One could then bind them easily to some key and everything would pretty much work for me again.

## Questions

1. Is there currently some nice way to trigger a specific autofill by calling some function?
2. Is there some way to fake a key press like the tab key?

## Implementation

Independent of what option we choose, I suggest to already implement four new functions:
1. `get-username-and-password`: this should prompt the user and return both username and password.
2. `get-username`: this should use `get-username-and-password` and return only the username.
3. `get-password`: this should use `get-username-and-password` and return only the password.
4. `get-password-key`: this should prompt the user for selecting a password entry and return it. I suggest that all functions triggering a password prompt use this, as currently I guess there is some duplicated code for `clip-username` and `clip-password`.

Note that `clip-username` and `clip-password` should probably use the new functions as well.

**Comment by aartaka** [comment]
Yes, those are all valid points. The refactoring point also makes lots of sense!

**Comment by aartaka** [comment]
@fabian-thomas answering your questions:
1. You can save an autofill to a variable and then invoke it with whatever the default autofills work with: `%paste`, for example. However, if would be nicer to make autofills funcallable and abstract their default action away :thinking: 
2. You can fake a keypress with `ffi-generate-input-event`, but it uses the underdocumented GTK-specific format for events that I am not even certain of how to use xD Definitely a space for improvement. For this particular case, we can focus the password element after the username one using JS.

**Comment by fabian-thomas** [comment]
Reading my above approach 2 again, I think nyxt should definitely provide some way to do that. So autofilling usernames into the username tagged fields. There are some conventions how those elements should be tagged. What do you think?

Regarding your point 2, I'm not that sure if focussing via JS would be the most robust solution, taking into account how broken the web is 😅 

Thanks for your answer.

**Comment by drone-ah** [comment]
Hi, I just discovered Nyxt, and as a vimmer, love the concept. Without a few more QoL, though, it's a bit tricky to switch over completely.

This issue re: username/password autofills is one such blocker.

What is the current status of this, specifically with Bitwarden?

While I am an engineer, I have zero experience in lisp, or functional languages. I would like to support, though, if I can - and would be willing to learn if it was a small task. But if it was a small task, it'd probably be done already I'm guessing!

**Comment by jmercouris** [comment]
Development is primarily focused on creating a new backend and fixing up old issues at this point. It will be some time yet until we can focus on adding something like BitWarden support.
